File: 2019/15/solution.ts
import * as fs from "fs";
import * as path from "path";
import intcodeGenerator from "../intcode";

const puzzleInput = fs
  .readFileSync(path.join(__dirname, "input.txt"), "utf-8")
  .split(",")
  .map(Number);

const visited: Set<string> = new Set(["0,0"]);
const queue: [number[], number[]][] = [[[], [0, 0]]];

const cloneRC = (commands: number[]) => {
  const newRC = intcodeGenerator(puzzleInput);
  for (const n of commands) {
    newRC.next();
    newRC.next(n);
  }
  return newRC;
};

const directions = [
  [0, 1],
  [0, -1],
  [1, 0],
  [-1, 0],
];

const areaMap: Map<string, number> = new Map([["0,0", 1]]);
let oxygenLocation: number[] = [0, 0];

let i = 0;
while (i < queue.length) {
  const [commands, pos] = queue[i];
  i++;
  for (let j = 1; j < 5; j++) {
    const [dx, dy] = directions[j - 1];
    const pos2 = [pos[0] + dx, pos[1] + dy];
    if (visited.has(pos2.join(","))) continue;
    visited.add(pos2.join(","));
    const remoteControl2 = cloneRC(commands);
    remoteControl2.next();
    const ret = remoteControl2.next(j);
    areaMap.set(pos2.join(","), ret.value);
    if (ret.value === 0) {
      continue;
    }
    if (ret.value === 2) {
      console.log(commands.length + 1);
      oxygenLocation = pos2;
    }
    queue.push([commands.concat(j), pos2]);
  }
}

const visited2: Set<string> = new Set([oxygenLocation.join(",")]);
const queue2: [number[], number][] = [[oxygenLocation, 0]];

let part2 = 0;
while (queue2.length) {
  const [pos, steps] = queue2.pop()!;
  part2 = Math.max(part2, steps);
  for (const [dx, dy] of directions) {
    const pos2 = [pos[0] + dx, pos[1] + dy];
    const coord2 = pos2.join(",");
    if (areaMap.get(coord2)! === 0) continue;
    if (visited2.has(coord2)) continue;
    visited2.add(coord2);
    queue2.push([pos2, steps + 1]);
  }
}

console.log(part2);

File: 2019/15/solution.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"slices"
	"strconv"
	"strings"

	"github.com/weitecklee/adventofcode/2019/intcode"
)

var directions = [][2]int{{0, 1}, {0, -1}, {-1, 0}, {1, 0}}

func main() {
	_, filename, _, _ := runtime.Caller(0)
	dirname := filepath.Dir(filename)
	inputFilePath := filepath.Join(dirname, "input.txt")
	data, err := os.ReadFile(inputFilePath)
	if err != nil {
		panic(err)
	}
	puzzleInput := parseInput(strings.Split(string(data), ","))
	fmt.Println(solve(puzzleInput))
}

func parseInput(data []string) []int {
	numbers := make([]int, 0, len(data))
	for _, s := range data {
		if n, err := strconv.Atoi(s); err != nil {
			panic(err)
		} else {
			numbers = append(numbers, n)
		}
	}
	return numbers
}

type QueueEntry struct {
	pos        [2]int
	ch         chan int
	cmdHistory []int
}

type QueueEntry2 struct {
	pos   [2]int
	steps int
}

func cloneDroid(puzzleInput, cmdHistory []int) chan int {
	ch := make(chan int)
	droid := intcode.NewIntcodeProgram(puzzleInput, ch)
	go droid.Run()
	for _, cmd := range cmdHistory {
		<-ch
		ch <- cmd
		<-ch
	}
	return ch
}

func reverse(dirIndex int) int {
	if dirIndex < 3 {
		return 3 - dirIndex
	}
	return 7 - dirIndex
}

func solve(puzzleInput []int) (int, int) {
	var part1, part2 int
	ch := make(chan int)
	droid := intcode.NewIntcodeProgram(puzzleInput, ch)
	go droid.Run()

	var queue []QueueEntry
	queue = append(queue, QueueEntry{[2]int{0, 0}, ch, []int{}})
	visited := make(map[[2]int]int)
	visited[[2]int{0, 0}] = 1

	var oxygenPos [2]int

	for len(queue) > 0 {
		item := queue[0]
		queue = queue[1:]
		pos := item.pos
		ch := item.ch
		cmdHistory := item.cmdHistory
		candidates := make([]int, 0, 4)

		for i, dir := range directions {
			pos2 := [2]int{pos[0] + dir[0], pos[1] + dir[1]}
			if _, ok := visited[pos2]; ok {
				continue
			}
			<-ch
			ch <- i + 1
			resp := <-ch
			visited[pos2] = resp
			if resp == 1 {
				candidates = append(candidates, i)
				<-ch
				ch <- reverse(i + 1)
				<-ch
			} else if resp == 2 {
				oxygenPos = pos2
				part1 = len(cmdHistory) + 1
			}
		}

		if len(candidates) == 0 {
			<-ch
			ch <- intcode.STOPSIGNAL
			continue
		}

		for _, i := range candidates[1:] {
			pos2 := [2]int{pos[0] + directions[i][0], pos[1] + directions[i][1]}
			ch2 := cloneDroid(puzzleInput, cmdHistory)
			<-ch2
			ch2 <- i + 1
			<-ch2
			queue = append(queue, QueueEntry{pos2, ch2, slices.Concat(cmdHistory, []int{i + 1})})
		}
		i := candidates[0]
		<-ch
		ch <- i + 1
		<-ch
		pos2 := [2]int{pos[0] + directions[i][0], pos[1] + directions[i][1]}
		queue = append(queue, QueueEntry{pos2, ch, append(cmdHistory, i+1)})
	}

	queue2 := []QueueEntry2{{oxygenPos, 0}}

	for len(queue2) > 0 {
		item := queue2[0]
		queue2 = queue2[1:]
		pos := item.pos
		steps := item.steps
		if steps > part2 {
			part2 = steps
		}
		for _, dir := range directions {
			pos2 := [2]int{pos[0] + dir[0], pos[1] + dir[1]}
			if visited[pos2] != 0 {
				visited[pos2] = 0
				queue2 = append(queue2, QueueEntry2{pos2, steps + 1})
			}
		}
	}

	return part1, part2
}
